### PR TITLE
Split kast module into kinner and kouter

### PR DESCRIPTION
**Pull request by ehildenb**
This makes a refactoring I've wanted for a while to organize the kast and kastManip modules a bit.

- New subdirectory `src/pyk/kast`, which contains modules:
  - `kinner.py`: part of what used to be `src/pyk/kast.py` (`KInner` datastructures)
  - `kouter.py`: part of what used to be `src/pyk/kast.py` (`KOuter` datastructures)
  - `manip.py`: what used to be `src/pyk/kastManip.py`.
- Adjust imports throughout the repository.
- In our `k-files`, adjust the `functional` attribute to the new `total` attribute instead.

### CODE REVIEW

File: src/pyk/kast/kinner.py [hunk 1]
```diff
@@ -0,0 +1,797 @@
...
+
+_LOGGER: Final = logging.getLogger(__name__)
+
+
+class KAst(ABC):
+    @classmethod
+    @abstractmethod
+    def from_dict(cls: Type[T], d: Dict[str, Any]) -> T:
+        node = d['node']
+        return globals()[node].from_dict(d)
```
Comment (tothtamas28) [inline src/pyk/kast/kinner.py:42]: This separation `KInner` and `KOuter` breaks `from_dict`, e.g.

```python3
KAst.from_dict({
    'node': 'KRequire',
    'require': 'imp.k',
})
```

will now fail. I tried to do this before, but couldn't find a solution (https://github.com/runtimeverification/pyk/issues/45).

Comment (ehildenb) [reply src/pyk/kast/kinner.py:42]: Yeah, but we never need to do `KAst.from_dict`, we only need to ever be able to do `KInner.from_dict` or `KOuter.from_dict`, but usually we even know more specifically which one we want (a few such adjustments are made in this PR, for example `KFlatModuleList.from_dict` or `KDefinition.from_dict`).

Any time you do `KAst.from_dict`, you immediately cast it to the more specific one you want anyway (so type-checking works), so you might as well just say `MORE_SPECIFIC.from_dict` anyway.

File: src/integration_tests/test_emit_json_spec.py [hunk 2]
```diff
@@ -27,8 +28,7 @@ def setUp(self) -> None:
         _kprove(spec_file, kompiled_dir=self.kompiled_dir, emit_json_spec=self.spec_json_file, dry_run=True)
 
         with open(self.spec_json_file, 'r') as f:
-            kfml = KAst.from_dict(json.load(f)['term'])
-            assert type(kfml) is KFlatModuleList
+            kfml = KFlatModuleList.from_dict(json.load(f)['term'])
```
Comment (ehildenb) [inline src/integration_tests/test_emit_json_spec.py:32]: This is one such example where we know we'll always be reading a `KFlatModuleList`, and we were casting it to that one anyway.

File: src/pyk/kast/kouter.py [hunk 3]
```diff
@@ -1849,12 +1145,6 @@ def build_cons(unit: KInner, label: Union[str, KLabel], terms: Iterable[KInner])
...
-    with open(path, 'r') as _f:
-        return KAst.from_dict(json.loads(_f.read())['term'])
-
-
 def read_kast_definition(path: Union[str, PathLike]) -> KDefinition:
-    _defn = read_kast(path)
-    assert isinstance(_defn, KDefinition)
-    return _defn
+    with open(path, 'r') as _f:
+        return KDefinition.from_dict(json.loads(_f.read())['term'])
```
Comment (ehildenb) [inline src/pyk/kast/kouter.py:1150]: Here we get rid of the generic `read_kast` which does not work with the new type structure (this is the only function that broke). But `read_kast` only use was in `read_kast_definition`, at least here and in KEVM, and `read_kast_definition` can just use the more specific `KDefinition.from_dict`.

File: src/pyk/kast/kinner.py (hunk 1 above)
Comment (tothtamas28) [reply src/pyk/kast/kinner.py:42]: Then `from_dict` and `from_json` need to be removed from `KAst`.

Comment (ehildenb) [reply src/pyk/kast/kinner.py:42]: or made virtual

Comment (tothtamas28) [reply src/pyk/kast/kinner.py:42]: You're right, replacing the body of `from_dict` with `...` makes more sense. Let's do that instead.

Comment (ehildenb) [reply src/pyk/kast/kinner.py]: Done.

File: src/pyk/kast/kinner.py [hunk 4]
```diff
@@ -0,0 +1,797 @@
...
+from ..utils import EMPTY_FROZEN_DICT, FrozenDict, hash_str
+
+T = TypeVar('T', bound='KAst')
+W = TypeVar('W', bound='WithKAtt')
+KI = TypeVar('KI', bound='KInner')
+
+_LOGGER: Final = logging.getLogger(__name__)
+
+
+class KAst(ABC):
```
[the remark this answers is not in the document]
Comment (ehildenb) [reply src/pyk/kast/kinner.py:37]: Note that using the same KOuter datastructures means that all the methods in `KDefinition` (for example) will have to be made over a new abstract `KInner2`, which is a `Union[KInner,KoreInner]`. Then queries like `KDefinition.rules` will work over both KAst and Kore.

**Review by tothtamas28 — APPROVED** [review]

File: src/integration_tests/kompiled_test.py [hunk 5]
```diff
@@ -5,7 +5,8 @@
 from typing import Iterable, Optional
 from unittest import TestCase
 
-from pyk.kast import KDefinition, KInner
+from pyk.kast.kinner import KInner
+from pyk.kast.kouter import KDefinition
```
Comment (tothtamas28) [inline src/integration_tests/kompiled_test.py:9]: What's your opinion about calling these modules `pyk.kast.inner` and `pyk.kast.outer` instead?

File: src/pyk/ktool/krun.py [hunk 6]
```diff
@@ -50,8 +50,7 @@ def run(
         if result.returncode != 0:
             raise RuntimeError('Non-zero exit-code from krun.')
 
-        result_kast = KAst.from_dict(json.loads(result.stdout)['term'])
-        assert isinstance(result_kast, KInner)
+        result_kast = KInner.from_dict(json.loads(result.stdout)['term'])
```
[the remark this answers is not in the document]
Comment (ehildenb) [reply src/pyk/ktool/krun.py:53]: But why? Do you have a use case for it? Issue created: https://github.com/runtimeverification/pyk/issues/97

File: src/integration_tests/kompiled_test.py (hunk 5 above)
Comment (ehildenb) [reply src/integration_tests/kompiled_test.py:9]: Sounds good, can do that.

**Comment by ehildenb** [comment]
I'll note that this better reflects the way that the Java handles these classes too.

**Pull request merged**